### PR TITLE
Focus management changes

### PR DESCRIPTION
**Pull request by yschimke**
#### WHAT

Add current version of FocusNode from https://android.googlesource.com/platform/frameworks/support/+/refs/changes/24/2171324/6/wear/compose/compose-foundation/src/commonMain/kotlin/androidx/wear/compose/foundation/Focus.kt

Replace all passing around of FocusRequester with this new approach

#### WHY

Testing the new approach and confirming it works.

#### Checklist :clipboard:
- [x] Add explicit visibility modifier and explicit return types for public declarations
- [x] Run spotless check
- [x] Run tests
- [x] Update metalava's signature text files

### CODE REVIEW

File: compose-layout/src/main/java/com/google/android/horologist/compose/navscaffold/WearNavScaffold.kt [hunk 1]
```diff
@@ -206,13 +207,17 @@ public fun NavGraphBuilder.scalingLazyColumnComposable(
     content: @Composable (ScaffoldContext<ScalingLazyListState>) -> Unit
 ) {
     composable(route, arguments, deepLinks) {
-        val viewModel: NavScaffoldViewModel = viewModel(it)
-
-        val scrollState = viewModel.initializeScalingLazyListState(scrollStateBuilder)
+        val lifecycle = LocalLifecycleOwner.current.lifecycle
```
Comment (yschimke) [inline compose-layout/src/main/java/com/google/android/horologist/compose/navscaffold/WearNavScaffold.kt:210]: This is a bug, lifecycle changes aren't triggering state updates.

File: compose-layout/src/main/java/com/google/android/horologist/compose/navscaffold/WearNavScaffold.kt [hunk 2]
```diff
@@ -274,25 +279,34 @@ public fun NavGraphBuilder.wearNavComposable(
...
-    // Wire up to NavBackStackEntry lifecycle
-    // events to make sure this composable handles
-    // events like scrolling.
-    LaunchedEffect(Unit) {
-        this@ResumeAsNeeded.repeatOnLifecycle(state = Lifecycle.State.STARTED) {
-            viewModel.resumed()
+internal fun FocusedDestination(route: String, content: @Composable () -> Unit) {
+    val lifecycle = LocalLifecycleOwner.current.lifecycle
+    val focused =
+        remember { mutableStateOf(lifecycle.currentState.isAtLeast(Lifecycle.State.RESUMED)) }
```
Comment (ssancho14) [inline compose-layout/src/main/java/com/google/android/horologist/compose/navscaffold/WearNavScaffold.kt:294]: You need an observer bellow?
Haven't used `LocalLifecycleOwner`, but in compose I'll expect the current state to be a `State`, and something like:
```
val focused =
        remember { derivedStateOf(lifecycle.currentState.isAtLeast(Lifecycle.State.RESUMED)) }
```
To be automatically registered an updated as needed.

**Review by ssancho14 — APPROVED** [review]

File: compose-layout/src/main/java/com/google/android/horologist/compose/pager/PagerScreen.kt [hunk 3]
```diff
@@ -64,47 +59,25 @@ public fun PagerScreen(
...
                     }
                 }
-                scopes[page] = scope
-                onDispose {
-                    coroutineScope.launch {
-                        scope.lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY)
+            ) {
+                FocusControl(requiresFocus = { page == state.currentPage }) {
+                    OnFocusChange {
+                        println("focus $page $it")
```
Comment (luizgrp) [inline compose-layout/src/main/java/com/google/android/horologist/compose/pager/PagerScreen.kt:79]: do you intend to merge the println?

File: compose-layout/src/main/java/com/google/android/horologist/compose/navscaffold/WearNavScaffold.kt [hunk 4]
```diff
@@ -274,25 +279,34 @@ public fun NavGraphBuilder.wearNavComposable(
...
+        }
+        lifecycle.addObserver(listener)
+        onDispose {
+            lifecycle.removeObserver(listener)
+        }
+    }
+
+    FocusControl(requiresFocus = { focused.value }) {
+        OnFocusChange {
+            println("focus $route $it")
```
Comment (luizgrp) [inline compose-layout/src/main/java/com/google/android/horologist/compose/navscaffold/WearNavScaffold.kt:308]: do you intend to merge the println?

File: compose-layout/src/main/java/com/google/android/horologist/compose/focus/FocusNode.kt [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+ * focus when this component is active.
+ * Note that this will may call requestFocus in the provided FocusRequester, so that focusRequester
+ * should be used in a .focusRequester modifier.
+ *
+ * @param focusRequester The associated [FocusRequester] to request focus on.
+ */
+@Composable
+public fun RequestFocusWhenActive(focusRequester: FocusRequester) {
+    OnFocusChange {
+        println("RequestFocusWhenActive $focusRequester $it")
```
Comment (luizgrp) [inline compose-layout/src/main/java/com/google/android/horologist/compose/focus/FocusNode.kt:93]: do you intend to merge the println?

**Review by luizgrp — APPROVED** [review]
nice improvement!

File: compose-layout/src/main/java/com/google/android/horologist/compose/pager/PagerScreen.kt (hunk 3 above)
Comment (yschimke) [reply compose-layout/src/main/java/com/google/android/horologist/compose/pager/PagerScreen.kt:79]: Oops!

File: compose-layout/src/main/java/com/google/android/horologist/compose/navscaffold/WearNavScaffold.kt (hunk 2 above)
Comment (yschimke) [reply compose-layout/src/main/java/com/google/android/horologist/compose/navscaffold/WearNavScaffold.kt:294]: Lifecycle is not compose aware.

**Pull request merged**